dev/source -> stable cutover can leave service wrapper and plugin runtime state inconsistent

## Summary

During a local dev/source -> stable package cutover to OpenClaw `2026.5.2` on macOS arm64, the gateway became partially inconsistent across service wrapper state, bundled/plugin channel loading, and plugin install metadata. The machine recovered after local manual repair, but the failure mode suggests the updater/cutover path can leave a packaged install in a state where the gateway runs while critical plugins/channels/context engine are unavailable or mis-resolved.

This is an evidence-backed incident report rather than a clean reproducer. I am filing it because the same system has also seen recent bundled/runtime-deps channel issues, and this cutover path may be adjacent to those fixes.

## Environment

- OpenClaw after recovery: `2026.5.2 (8b2a6e5)`
- Install root after recovery: `/opt/homebrew/lib/node_modules/openclaw`
- Install kind after recovery: package
- OS: macOS arm64
- Gateway: local LaunchAgent, loopback gateway
- Main channel affected: Discord
- Context engine plugin affected: `@martian-engineering/lossless-claw`

## Observed symptoms

1. The service env wrapper path became corrupted during the cutover. Shell history preserved a manual recovery command copying from a literal markdown-link-style filename back to the expected wrapper filename:

   ```text
   cp ~/.openclaw/service-env/'[ai.openclaw.gateway-env-wrapper.sh](http://ai.openclaw.gateway-env-wrapper.sh)' ~/.openclaw/service-env/ai.openclaw.gateway-env-wrapper.sh
   ```

   After recovery, the LaunchAgent points to the corrected wrapper path and the service-env directory contains only the expected wrapper/env files.

2. Discord plugin/channel loading failed repeatedly before recovery with:

   ```text
   [channels] failed to load bundled channel discord: plugin module path escapes plugin root or fails alias checks
   ```

   This appeared multiple times during the cutover window.

3. The context engine fell back because Lossless Claw was not registered:

   ```text
   [context-engine] Context engine "lossless-claw" is not registered; falling back to default engine "legacy"
   ```

4. Lossless Claw plugin install metadata drifted and required repinning/recovery to the canonical npm package:

   ```text
   @martian-engineering/lossless-claw@0.9.2
   ```

   Gateway logs recorded a restart request after updating plugin load path/install metadata for the pinned package.

5. Node-side runtime import mismatch appeared once:

   ```text
   Error [ERR_MODULE_NOT_FOUND]: Cannot find module '/opt/homebrew/lib/node_modules/openclaw/dist/config-X3ZUqr4X.js' imported from /opt/homebrew/lib/node_modules/openclaw/dist/node-cli-DzSJTG00.js
   ```

## Recovery evidence

After manual repair/restart, the gateway recovered cleanly:

```text
[gateway] http server listening (3 plugins: browser, discord, memory-core; 3.6s)
[discord] starting provider
```

Current status confirms:

```text
OpenClaw 2026.5.2 (8b2a6e5)
runtimeVersion: 2026.5.2
install root: /opt/homebrew/lib/node_modules/openclaw
install kind: package
gateway reachable: true
```

Lossless Claw is now pinned at:

```text
~/.openclaw/npm/node_modules/@martian-engineering/lossless-claw/package.json
name: @martian-engineering/lossless-claw
version: 0.9.2
```

## Why this may matter

The concerning part is not any single plugin failure; it is that a dev/source -> stable package cutover appeared to leave several runtime surfaces inconsistent at the same time:

- service wrapper filename/path state
- bundled/channel plugin root or alias checks
- plugin install metadata / load path
- context engine registration
- node CLI module import path

The gateway can appear partially alive while the operator-facing channel and context engine are degraded.

## Reproduction status

I have not attempted to reproduce this on the live machine because re-entering the dev/source -> stable cutover path risks taking the gateway/channel offline again. The local evidence has been preserved and independently verified against logs.

## Expected behavior

A dev/source -> stable package cutover should either:

- complete atomically with service wrapper, package root, plugin installs, bundled channel deps, and plugin load paths consistent; or
- fail closed with a clear repair instruction rather than leaving the gateway partially running with missing channel/context functionality.

## Related issues possibly adjacent

This may be adjacent to recent bundled/runtime dependency reports, but it is not the same exact symptom as missing a single package dependency. It includes service wrapper corruption and plugin install/load-path drift during cutover.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -35,6 +35,7 @@ Docs: https://docs.openclaw.ai
 
 ### Fixes
 
+- CLI/update: preserve managed Gateway service environment during package cutovers so macOS LaunchAgent repair/restart reads the pre-update service state instead of caller shell state. (#83026)
 - Gateway/restart: eager-load the lifecycle runtime before in-place upgrade signal handling so package replacement does not deadlock restart imports. (#84890) Thanks @myps6415.
 - CLI/update: start managed Gateway update handoff helpers from a stable existing directory and tolerate deleted cwd/package roots during macOS LaunchAgent handoff. Fixes #83808. (#83875) Thanks @jason-allen-oneal.
 - Cron: honor `cron.retry.retryOn: ["network"]` for common network error codes such as `EAI_AGAIN`, `EHOSTUNREACH`, and `ENETUNREACH`.

--- a/src/cli/update-cli/update-command.test.ts
+++ b/src/cli/update-cli/update-command.test.ts
@@ -14,6 +14,7 @@ import {
   recoverInstalledLaunchAgentAfterUpdate,
   recoverLaunchAgentAndRecheckGatewayHealth,
   resolvePostCoreUpdateChildStdio,
+  resolvePostUpdateServiceStateReadEnv,
   resolvePostInstallDoctorEnv,
   shouldPrepareUpdatedInstallRestart,
   resolveUpdatedGatewayRestartPort,
@@ -114,6 +115,40 @@ describe("resolveUpdatedGatewayRestartPort", () => {
         serviceEnv: {},
       }),
     ).toBe(19000);
+  });
+});
+
+describe("resolvePostUpdateServiceStateReadEnv", () => {
+  it("keeps package restart preparation anchored to the pre-update service env", () => {
+    const processEnv = {
+      OPENCLAW_STATE_DIR: "/source/state",
+      OPENCLAW_CONFIG_PATH: "/source/openclaw.json",
+    } as NodeJS.ProcessEnv;
+    const prePackageServiceEnv = {
+      OPENCLAW_STATE_DIR: "/managed/state",
+      OPENCLAW_CONFIG_PATH: "/managed/openclaw.json",
+    } as NodeJS.ProcessEnv;
+
+    expect(
+      resolvePostUpdateServiceStateReadEnv({
+        updateMode: "npm",
+        processEnv,
+        prePackageServiceEnv,
+      }),
+    ).toBe(prePackageServiceEnv);
+  });
+
+  it("keeps git updates tied to the caller environment", () => {
+    const processEnv = { OPENCLAW_STATE_DIR: "/source/state" } as NodeJS.ProcessEnv;
+    const prePackageServiceEnv = { OPENCLAW_STATE_DIR: "/managed/state" } as NodeJS.ProcessEnv;
+
+    expect(
+      resolvePostUpdateServiceStateReadEnv({
+        updateMode: "git",
+        processEnv,
+        prePackageServiceEnv,
+      }),
+    ).toBe(processEnv);
   });
 });
 

--- a/src/cli/update-cli/update-command.ts
+++ b/src/cli/update-cli/update-command.ts
@@ -1086,6 +1086,17 @@ export function resolveUpdatedGatewayRestartPort(params: {
   return resolveGatewayPort(params.config, params.serviceEnv ?? params.processEnv ?? process.env);
 }
 
+export function resolvePostUpdateServiceStateReadEnv(params: {
+  updateMode: UpdateRunResult["mode"];
+  processEnv?: NodeJS.ProcessEnv;
+  prePackageServiceEnv?: NodeJS.ProcessEnv;
+}): NodeJS.ProcessEnv {
+  if (isPackageManagerUpdateMode(params.updateMode) && params.prePackageServiceEnv) {
+    return params.prePackageServiceEnv;
+  }
+  return params.processEnv ?? process.env;
+}
+
 type UpdateDryRunPreview = {
   dryRun: true;
   root: string;
@@ -3502,7 +3513,11 @@ export async function updateCommand(opts: UpdateCommandOptions): Promise<void> {
   if (shouldRestart) {
     try {
       const serviceState = await readGatewayServiceState(resolveGatewayService(), {
-        env: process.env,
+        env: resolvePostUpdateServiceStateReadEnv({
+          updateMode: resultWithPostUpdate.mode,
+          processEnv: process.env,
+          prePackageServiceEnv: prePackageServiceStop?.serviceEnv,
+        }),
       });
       if (
         shouldPrepareUpdatedInstallRestart({

--- a/src/daemon/launchd-plist.ts
+++ b/src/daemon/launchd-plist.ts
@@ -27,6 +27,12 @@ const plistUnescape = (value: string): string =>
     .replaceAll("&lt;", "<")
     .replaceAll("&amp;", "&");
 
+type ReadLaunchAgentProgramArgumentsOptions = {
+  expectedEnvironmentWrapperPath?: string;
+  expectedEnvironmentFilePath?: string;
+  generatedEnvironmentLabel?: string;
+};
+
 function parseGeneratedEnvValue(value: string): string {
   const trimmed = value.trim();
   if (!trimmed.startsWith("'") || !trimmed.endsWith("'")) {
@@ -35,17 +41,93 @@ function parseGeneratedEnvValue(value: string): string {
   return trimmed.slice(1, -1).replaceAll("'\\''", "'");
 }
 
+function includesGeneratedEnvironmentPathToken(value: string | undefined, token: string): boolean {
+  return Boolean(value?.replaceAll("\\", "/").includes(token));
+}
+
+function includesGeneratedEnvironmentDirToken(value: string | undefined): boolean {
+  return Boolean(value?.replaceAll("\\", "/").includes("/service-env/"));
+}
+
+function resolveSiblingGeneratedEnvFilePath(
+  envFilePath: string,
+  options?: ReadLaunchAgentProgramArgumentsOptions,
+): string | undefined {
+  const label = options?.generatedEnvironmentLabel?.trim();
+  if (!label) {
+    return undefined;
+  }
+  const serviceEnvMarker = "/service-env/";
+  const markerIndex = envFilePath.replaceAll("\\", "/").lastIndexOf(serviceEnvMarker);
+  if (markerIndex < 0) {
+    return undefined;
+  }
+  // Custom state dirs can also contain service-env; use the generated env dir closest to the file.
+  const serviceEnvDirEnd = markerIndex + serviceEnvMarker.length - 1;
+  return `${envFilePath.slice(0, serviceEnvDirEnd)}/${label}.env`;
+}
+
+function isGeneratedEnvWrapperArgs(
+  programArguments: string[],
+  options?: ReadLaunchAgentProgramArgumentsOptions,
+): boolean {
+  const wrapperPath = programArguments[0];
+  const envFilePath = programArguments[1];
+  if (!wrapperPath || !envFilePath) {
+    return false;
+  }
+  if (!options) {
+    return wrapperPath.endsWith("-env-wrapper.sh");
+  }
+  if (
+    options.expectedEnvironmentWrapperPath &&
+    options.expectedEnvironmentFilePath &&
+    wrapperPath === options.expectedEnvironmentWrapperPath &&
+    envFilePath === options.expectedEnvironmentFilePath
+  ) {
+    return true;
+  }
+  const label = options.generatedEnvironmentLabel?.trim();
+  if (!label) {
+    return false;
+  }
+  // Legacy/corrupted plists may preserve the label-derived wrapper name inside
+  // a mangled service-env path. Still unwrap it so the next rewrite can repair.
+  return (
+    includesGeneratedEnvironmentDirToken(wrapperPath) &&
+    includesGeneratedEnvironmentDirToken(envFilePath) &&
+    includesGeneratedEnvironmentPathToken(wrapperPath, `${label}-env-wrapper.sh`) &&
+    includesGeneratedEnvironmentPathToken(envFilePath, `${label}.env`)
+  );
+}
+
 async function readLaunchAgentEnvironmentFile(
   programArguments: string[],
+  options?: ReadLaunchAgentProgramArgumentsOptions,
 ): Promise<Record<string, string>> {
   const envFilePath = programArguments[1];
-  if (!programArguments[0]?.endsWith("-env-wrapper.sh") || !envFilePath) {
+  if (!isGeneratedEnvWrapperArgs(programArguments, options) || !envFilePath) {
     return {};
   }
   let content = "";
-  try {
-    content = await fs.readFile(envFilePath, "utf8");
-  } catch {
+  const candidateEnvFilePaths = Array.from(
+    new Set(
+      [
+        envFilePath,
+        resolveSiblingGeneratedEnvFilePath(envFilePath, options),
+        options?.expectedEnvironmentFilePath,
+      ].filter((candidate): candidate is string => Boolean(candidate)),
+    ),
+  );
+  for (const candidate of candidateEnvFilePaths) {
+    try {
+      content = await fs.readFile(candidate, "utf8");
+      break;
+    } catch {
+      // Keep trying; mangled wrapper args may still have the canonical env file.
+    }
+  }
+  if (!content) {
     return {};
   }
   const environment: Record<string, string> = {};
@@ -68,8 +150,11 @@ async function readLaunchAgentEnvironmentFile(
   return environment;
 }
 
-function unwrapGeneratedEnvWrapperArgs(programArguments: string[]): string[] {
-  if (!programArguments[0]?.endsWith("-env-wrapper.sh") || !programArguments[1]) {
+function unwrapGeneratedEnvWrapperArgs(
+  programArguments: string[],
+  options?: ReadLaunchAgentProgramArgumentsOptions,
+): string[] {
+  if (!isGeneratedEnvWrapperArgs(programArguments, options)) {
     return programArguments;
   }
   return programArguments.slice(2);
@@ -95,6 +180,26 @@ const renderEnvDict = (env: Record<string, string | undefined> | undefined): str
 };
 
 export async function readLaunchAgentProgramArgumentsFromFile(plistPath: string): Promise<{
+  programArguments: string[];
+  workingDirectory?: string;
+  environment?: Record<string, string>;
+  environmentValueSources?: Record<string, GatewayServiceEnvironmentValueSource>;
+  sourcePath?: string;
+} | null>;
+export async function readLaunchAgentProgramArgumentsFromFile(
+  plistPath: string,
+  options: ReadLaunchAgentProgramArgumentsOptions,
+): Promise<{
+  programArguments: string[];
+  workingDirectory?: string;
+  environment?: Record<string, string>;
+  environmentValueSources?: Record<string, GatewayServiceEnvironmentValueSource>;
+  sourcePath?: string;
+} | null>;
+export async function readLaunchAgentProgramArgumentsFromFile(
+  plistPath: string,
+  options?: ReadLaunchAgentProgramArgumentsOptions,
+): Promise<{
   programArguments: string[];
   workingDirectory?: string;
   environment?: Record<string, string>;
@@ -128,8 +233,8 @@ export async function readLaunchAgentProgramArgumentsFromFile(plistPath: string)
         inlineEnvironment[key] = value;
       }
     }
-    const fileEnvironment = await readLaunchAgentEnvironmentFile(args);
-    const effectiveProgramArguments = unwrapGeneratedEnvWrapperArgs(args);
+    const fileEnvironment = await readLaunchAgentEnvironmentFile(args, options);
+    const effectiveProgramArguments = unwrapGeneratedEnvWrapperArgs(args, options);
     const environment = { ...inlineEnvironment, ...fileEnvironment };
     const environmentValueSources: Record<string, GatewayServiceEnvironmentValueSource> = {};
     for (const key of Object.keys(inlineEnvironment)) {

--- a/src/daemon/launchd.test.ts
+++ b/src/daemon/launchd.test.ts
@@ -726,6 +726,63 @@ describe("launchd install", () => {
     expect(command?.environmentValueSources?.OPENAI_API_KEY).toBe("file");
   });
 
+  it("repairs a mangled label-derived service-env wrapper path on restart", async () => {
+    const callerEnv = createDefaultLaunchdEnv();
+    const serviceEnv = {
+      ...callerEnv,
+      OPENCLAW_STATE_DIR: "/Users/test/service-env/custom-state",
+    };
+    await installLaunchAgent({
+      env: serviceEnv,
+      stdout: new PassThrough(),
+      programArguments: defaultProgramArguments,
+      environment: {
+        OPENCLAW_GATEWAY_PORT: "18789",
+        OPENCLAW_STATE_DIR: serviceEnv.OPENCLAW_STATE_DIR,
+      },
+    });
+
+    const plistPath = resolveLaunchAgentPlistPath(callerEnv);
+    const envFilePath = "/Users/test/service-env/custom-state/service-env/ai.openclaw.gateway.env";
+    const wrapperPath =
+      "/Users/test/service-env/custom-state/service-env/ai.openclaw.gateway-env-wrapper.sh";
+    const callerEnvFilePath = "/Users/test/.openclaw/service-env/ai.openclaw.gateway.env";
+    const callerWrapperPath =
+      "/Users/test/.openclaw/service-env/ai.openclaw.gateway-env-wrapper.sh";
+    const mangledEnvFilePath =
+      "/Users/test/service-env/custom-state/service-env/[ai.openclaw.gateway.env](http:/ai.openclaw.gateway.env)";
+    const mangledWrapperPath =
+      "/Users/test/service-env/custom-state/service-env/[ai.openclaw.gateway-env-wrapper.sh](http:/ai.openclaw.gateway-env-wrapper.sh)";
+    state.files.set(
+      plistPath,
+      (state.files.get(plistPath) ?? "")
+        .replace(wrapperPath, mangledWrapperPath)
+        .replace(envFilePath, mangledEnvFilePath),
+    );
+
+    const command = await readLaunchAgentProgramArguments(callerEnv);
+    expect(command?.programArguments).toEqual(defaultProgramArguments);
+    expect(command?.environment?.OPENCLAW_GATEWAY_PORT).toBe("18789");
+    expect(command?.environment?.OPENCLAW_STATE_DIR).toBe(serviceEnv.OPENCLAW_STATE_DIR);
+    expect(command?.environmentValueSources?.OPENCLAW_GATEWAY_PORT).toBe("file");
+
+    await restartLaunchAgent({
+      env: callerEnv,
+      stdout: new PassThrough(),
+    });
+
+    const rewritten = state.files.get(plistPath) ?? "";
+    expect(rewritten).toContain(`<string>${callerWrapperPath}</string>`);
+    expect(rewritten).toContain(`<string>${callerEnvFilePath}</string>`);
+    expect(rewritten).not.toContain(mangledEnvFilePath);
+    expect(rewritten).not.toContain(mangledWrapperPath);
+    const rewrittenEnv = state.files.get(callerEnvFilePath) ?? "";
+    expect(rewrittenEnv).toContain("export OPENCLAW_GATEWAY_PORT='18789'");
+    expect(rewrittenEnv).toContain(
+      "export OPENCLAW_STATE_DIR='/Users/test/service-env/custom-state'",
+    );
+  });
+
   it("creates the LaunchAgent TMPDIR before bootstrap", async () => {
     const env = createDefaultLaunchdEnv();
     const tmpDir = "/Users/test/.openclaw/tmp";

--- a/src/daemon/launchd.ts
+++ b/src/daemon/launchd.ts
@@ -232,11 +232,23 @@ export function resolveLaunchAgentPlistPath(env: GatewayServiceEnv): string {
   return resolveLaunchAgentPlistPathForLabel(env, label);
 }
 
+function resolveLaunchAgentEnvironmentReadOptions(env: GatewayServiceEnv, label: string) {
+  return {
+    expectedEnvironmentWrapperPath: resolveLaunchAgentEnvWrapperPath(env, label),
+    expectedEnvironmentFilePath: resolveLaunchAgentEnvFilePath(env, label),
+    generatedEnvironmentLabel: label,
+  };
+}
+
 export async function readLaunchAgentProgramArguments(
   env: GatewayServiceEnv,
 ): Promise<GatewayServiceCommandConfig | null> {
+  const label = resolveLaunchAgentLabel({ env });
   const plistPath = resolveLaunchAgentPlistPath(env);
-  return readLaunchAgentProgramArgumentsFromFile(plistPath);
+  return readLaunchAgentProgramArgumentsFromFile(
+    plistPath,
+    resolveLaunchAgentEnvironmentReadOptions(env, label),
+  );
 }
 
 function buildLaunchAgentPlist({
@@ -926,7 +938,10 @@ async function rewriteLaunchAgentPlistForRestart({
   label: string;
   plistPath: string;
 }): Promise<boolean> {
-  const existing = await readLaunchAgentProgramArgumentsFromFile(plistPath);
+  const existing = await readLaunchAgentProgramArgumentsFromFile(
+    plistPath,
+    resolveLaunchAgentEnvironmentReadOptions(env, label),
+  );
   if (!existing?.programArguments.length) {
     return false;
   }

